doc/designs/quic-design/quic-record-layer.md

```diff
@@ -0,0 +1,101 @@
+Design Problem: Record Layer
+============================
+
+* BIO filter
+    * Later exposed via provider
+    * Ctrls
+* RECORD METHOD structure
+    * Later exposed via provider
+    * Could be individual callback mechanism
+    * PARAMS + arguments
+
+The record layer transmission is a simple layer which takes data from
+upper levels, adds protection and (possibly) record headers and passes it
+to the next layer for sending. Receiving packets is the opposite
+operation: data is received from the network layer, protection is removed
+and plaintext data is passed to the upper layers. The record layer API
+needs to be universal - it needs to work for TLS, DTLS, KTLS and QUIC.
+
+Definition of _single-copy_: Given there needs to happen an encryption
+or decryption to protect or unprotect the data send or received, a copy
+of the data (in the transformed form) is always created. Thus the
+usual term of _zero-copy_ does not make sense here and we use the term
+_single-copy_.
+
+Candidate Solution: BIO filter
+------------------------------
+
+We have an existing BIO abstraction. The record layer could be modelled
+as a filter BIO. To send packets the upper layers call `BIO_write()`. The
+record layer adds the protections and passes the data down to a further
+network layer - which could be another BIO in the chain.
+
+To receive packets the upper layers call `BIO_read()`. The record layer
+reads data from the network layer (possibly via a BIO), removes
+protection and fills the buffer supplied by the upper layers with the
+received data.
+
+Retransmits are performed by the upper layers. However they will need to
```

Comment:
(applies to both candidate solutions listed, but this one is first)
We say we are going to support at least "TLS, DTLS, KTLS and QUIC".
TLS presumes a reliable stream as the underlay, and needs no retransmission.
DTLS only implements a reliability mechanism for its own handshake, not for application data.  So in this case the "upper layer" would still be in libssl and we would control it fully.
KTLS, AFAIK, only currently supports non-D TLS and the same situation for TLS applies.
QUIC does do retransmission but IIRC the retransmission is done as entirely new frames and records, so the record layer doesn't need to know that they are retransmissions.

DTLS, of course, will differ significantly between 1.3 and pre-1.3 -- 1.3 ACKs individual records but 1.2 retransmits the entire flight.  Also, the `message_seq` is reused in 1.2 but 1.3 produces a new `message_seq` at the Handshake layer.  This is, however, distinct from the record layer, where both 1.3 and 1.2 will use a new `sequence_number` value, so deferring retransmission to the upper layer (from the perspective of the record layer) does still make sense.

That said, we might say "Any retransmission necessary" since in some cases they are not.